File: complexidade/javascript/binary-search.js
function binarySearch(elem, arr) {
  return binarySearchTailRecursive(elem, arr, 0, arr.length - 1);
}

function binarySearchTailRecursive(elem, arr, start, finish) {
  if (start > finish) {
    return -1;
  }
  
  const middle = Math.floor((start + finish) / 2);
  if (elem == arr[middle]) {
    return middle;
  }
  else if (elem < arr[middle]) {
    return binarySearchTailRecursive(elem, arr, start, middle - 1);
  }
  else {
    return binarySearchTailRecursive(elem, arr, middle + 1, finish);
  }
}

const v = [7, 13, 20, 25, 28, 31, 35, 39, 40, 45, 46, 48, 57, 59, 63, 71];
console.log(binarySearch(39, v));
console.log(binarySearch(46, v));
console.log(binarySearch(22, v));

File: complexidade/python/binary-search.py
def binary_search(elem, arr):
    return binary_search_tail_recursive(elem, arr, 0, len(arr) - 1)


def binary_search_tail_recursive(elem, arr, start, finish):
    if start > finish:
        return -1
    middle = (start + finish) // 2
    if elem == arr[middle]:
        return middle
    elif elem < arr[middle]:
        return binary_search_tail_recursive(elem, arr, start, middle - 1)
    else:
        return binary_search_tail_recursive(elem, arr, middle + 1, finish)


v = [7, 13, 20, 25, 28, 31, 35, 39, 40, 45, 46, 48, 57, 59, 63, 71]
print(binary_search(39, v))
print(binary_search(46, v))
print(binary_search(22, v))
